File: shell/friendly.c
#include "friendly.h"
#include "../kernel/lib/string.h"
#include "../kernel/lib/printf.h"
#include "../kernel/db/database.h"

/* ---- Table aliases ---- */
static const table_alias_t aliases[] = {
    { "procs",   "ProcessTable" },
    { "caps",    "CapabilityTable" },
    { "objects", "ObjectTable" },
    { "msgs",    "MessageTable" },
    { "audit",   "AuditTable" },
    { "config",  "SystemTable" },
    { "sys",     "SystemTable" },
    { NULL, NULL }
};

const table_alias_t *friendly_get_aliases(void) {
    return aliases;
}

const char *friendly_resolve_alias(const char *name) {
    for (int i = 0; aliases[i].alias; i++) {
        if (strcasecmp(name, aliases[i].alias) == 0)
            return aliases[i].table;
    }
    return name;
}

/* ---- Friendly verbs ---- */
static const char *verbs[] = {
    "tables", "show", "info", "find", "count", "add", "del", "set",
    "create", "open", "list", "rm", "cat", "ps", "scripts", "inbox",
    "save", "run", "kill", "spawn", "msg",
    NULL
};

bool friendly_is_verb(const char *word, size_t len) {
    for (int i = 0; verbs[i]; i++) {
        if (strlen(verbs[i]) == len && strncasecmp(verbs[i], word, len) == 0)
            return true;
    }
    return false;
}

/* ---- Parsing helpers ---- */

/* Skip whitespace, return pointer to next non-space char */
static const char *skip_spaces(const char *s) {
    while (*s && isspace(*s)) s++;
    return s;
}

/* Extract a token (word) from s into buf. Returns pointer past the token. */
static const char *next_token(const char *s, char *buf, size_t buf_size) {
    s = skip_spaces(s);
    size_t i = 0;
    while (*s && !isspace(*s) && i < buf_size - 1) {
        buf[i++] = *s++;
    }
    buf[i] = '\0';
    return s;
}

/* Check if a string looks numeric (digits only, optional leading minus) */
static bool is_numeric(const char *s) {
    if (!*s) return false;
    if (*s == '-') s++;
    if (!*s) return false;
    while (*s) {
        if (!isdigit(*s)) return false;
        s++;
    }
    return true;
}

/* Append a value to SQL, quoting strings but not numbers/booleans */
static size_t append_value(char *sql, size_t pos, size_t sql_size, const char *val) {
    if (is_numeric(val) ||
        strcasecmp(val, "true") == 0 ||
        strcasecmp(val, "false") == 0) {
        /* Numeric or boolean: no quotes */
        int n = snprintf(sql + pos, sql_size - pos, "%s", val);
        if (n > 0) pos += (size_t)n;
    } else {
        /* String: wrap in single quotes */
        int n = snprintf(sql + pos, sql_size - pos, "'%s'", val);
        if (n > 0) pos += (size_t)n;
    }
    return pos;
}

/* Parse "col=val" pair. Returns pointer past the pair, or NULL on failure. */
static const char *parse_pair(const char *s, char *col, size_t col_size,
                               char *val, size_t val_size) {
    s = skip_spaces(s);
    if (!*s) return NULL;

    /* Read column name (up to '=') */
    size_t i = 0;
    while (*s && *s != '=' && !isspace(*s) && i < col_size - 1) {
        col[i++] = *s++;
    }
    col[i] = '\0';
    if (i == 0 || *s != '=') return NULL;
    s++; /* skip '=' */

    /* Read value (up to next space or end) */
    i = 0;
    if (*s == '\'') {
        /* Quoted value */
        s++;
        while (*s && *s != '\'' && i < val_size - 1) {
            val[i++] = *s++;
        }
        if (*s == '\'') s++;
    } else {
        while (*s && !isspace(*s) && i < val_size - 1) {
            val[i++] = *s++;
        }
    }
    val[i] = '\0';
    return s;
}

/* ---- Main translator ---- */

bool friendly_translate(const char *input, char *sql_out, size_t sql_size) {
    char verb[32];
    const char *rest = next_token(input, verb, sizeof(verb));

    if (verb[0] == '\0') return false;

    /* tables */
    if (strcasecmp(verb, "tables") == 0) {
        snprintf(sql_out, sql_size, "SHOW TABLES");
        return true;
    }

    /* show <table> */
    if (strcasecmp(verb, "show") == 0) {
        char table[MAX_TABLE_NAME];
        rest = next_token(rest, table, sizeof(table));
        if (table[0] == '\0') return false;
        /* "show tables" is also valid */
        if (strcasecmp(table, "tables") == 0) {
            snprintf(sql_out, sql_size, "SHOW TABLES");
            return true;
        }
        const char *resolved = friendly_resolve_alias(table);
        snprintf(sql_out, sql_size, "SELECT * FROM %s", resolved);
        return true;
    }

    /* info <table> */
    if (strcasecmp(verb, "info") == 0) {
        char table[MAX_TABLE_NAME];
        rest = next_token(rest, table, sizeof(table));
        if (table[0] == '\0') return false;
        const char *resolved = friendly_resolve_alias(table);
        snprintf(sql_out, sql_size, "DESCRIBE %s", resolved);
        return true;
    }

    /* count <table> */
    if (strcasecmp(verb, "count") == 0) {
        char table[MAX_TABLE_NAME];
        rest = next_token(rest, table, sizeof(table));
        if (table[0] == '\0') return false;
        const char *resolved = friendly_resolve_alias(table);
        /* Use SELECT * — the shell will show row count in the result */
        snprintf(sql_out, sql_size, "SELECT * FROM %s", resolved);
        return true;
    }

    /* find <table> [col=val ...] */
    if (strcasecmp(verb, "find") == 0) {
        char table[MAX_TABLE_NAME];
        rest = next_token(rest, table, sizeof(table));
        if (table[0] == '\0') return false;
        const char *resolved = friendly_resolve_alias(table);

        /* Check for conditions */
        rest = skip_spaces(rest);
        if (*rest == '\0') {
            /* No conditions: show all */
            snprintf(sql_out, sql_size, "SELECT * FROM %s", resolved);
            return true;
        }

        size_t pos = (size_t)snprintf(sql_out, sql_size, "SELECT * FROM %s WHERE ", resolved);
        bool first = true;

        while (*rest) {
            char col[MAX_COLUMN_NAME], val[MAX_STR_LEN];
            const char *next = parse_pair(rest, col, sizeof(col), val, sizeof(val));
            if (!next) break;

            if (!first) {
                int n = snprintf(sql_out + pos, sql_size - pos, " AND ");
                if (n > 0) pos += (size_t)n;
            }
            int n = snprintf(sql_out + pos, sql_size - pos, "%s = ", col);
            if (n > 0) pos += (size_t)n;
            pos = append_value(sql_out, pos, sql_size, val);

            first = false;
            rest = next;
            rest = skip_spaces(rest);
        }
        return true;
    }

    /* add <table> col=val [col=val ...] */
    if (strcasecmp(verb, "add") == 0) {
        char table[MAX_TABLE_NAME];
        rest = next_token(rest, table, sizeof(table));
        if (table[0] == '\0') return false;
        const char *resolved = friendly_resolve_alias(table);

        /* Collect col=val pairs */
        char cols[MAX_INSERT_VALS][MAX_COLUMN_NAME];
        char vals[MAX_INSERT_VALS][MAX_STR_LEN];
        int pair_count = 0;

        rest = skip_spaces(rest);
        while (*rest && pair_count < MAX_INSERT_VALS) {
            const char *next = parse_pair(rest, cols[pair_count], MAX_COLUMN_NAME,
                                           vals[pair_count], MAX_STR_LEN);
            if (!next) break;
            pair_count++;
            rest = next;
            rest = skip_spaces(rest);
        }

        if (pair_count == 0) return false;

        /* Build INSERT */
        size_t pos = (size_t)snprintf(sql_out, sql_size, "INSERT INTO %s (", resolved);
        for (int i = 0; i < pair_count; i++) {
            int n = snprintf(sql_out + pos, sql_size - pos, "%s%s",
                             i > 0 ? ", " : "", cols[i]);
            if (n > 0) pos += (size_t)n;
        }
        {
            int n = snprintf(sql_out + pos, sql_size - pos, ") VALUES (");
            if (n > 0) pos += (size_t)n;
        }
        for (int i = 0; i < pair_count; i++) {
            if (i > 0) {
                int n = snprintf(sql_out + pos, sql_size - pos, ", ");
                if (n > 0) pos += (size_t)n;
            }
            pos = append_value(sql_out, pos, sql_size, vals[i]);
        }
        snprintf(sql_out + pos, sql_size - pos, ")");
        return true;
    }

    /* del <table> col=val [col=val ...] */
    if (strcasecmp(verb, "del") == 0) {
        char table[MAX_TABLE_NAME];
        rest = next_token(rest, table, sizeof(table));
        if (table[0] == '\0') return false;
        const char *resolved = friendly_resolve_alias(table);

        rest = skip_spaces(rest);
        if (*rest == '\0') return false; /* require at least one condition */

        size_t pos = (size_t)snprintf(sql_out, sql_size, "DELETE FROM %s WHERE ", resolved);
        bool first = true;

        while (*rest) {
            char col[MAX_COLUMN_NAME], val[MAX_STR_LEN];
            const char *next = parse_pair(rest, col, sizeof(col), val, sizeof(val));
            if (!next) break;

            if (!first) {
                int n = snprintf(sql_out + pos, sql_size - pos, " AND ");
                if (n > 0) pos += (size_t)n;
            }
            int n = snprintf(sql_out + pos, sql_size - pos, "%s = ", col);
            if (n > 0) pos += (size_t)n;
            pos = append_value(sql_out, pos, sql_size, val);

            first = false;
            rest = next;
            rest = skip_spaces(rest);
        }
        return true;
    }

    /* set <table> col=val [col=val] where key=kval */
    if (strcasecmp(verb, "set") == 0) {
        char table[MAX_TABLE_NAME];
        rest = next_token(rest, table, sizeof(table));
        if (table[0] == '\0') return false;
        const char *resolved = friendly_resolve_alias(table);

        /* Collect SET pairs until "where" keyword */
        char set_cols[MAX_INSERT_VALS][MAX_COLUMN_NAME];
        char set_vals[MAX_INSERT_VALS][MAX_STR_LEN];
        int set_count = 0;

        rest = skip_spaces(rest);
        while (*rest && set_count < MAX_INSERT_VALS) {
            /* Check for "where" keyword */
            char peek[32];
            const char *peek_rest = next_token(rest, peek, sizeof(peek));
            if (strcasecmp(peek, "where") == 0) {
                rest = peek_rest;
                break;
            }

            const char *next = parse_pair(rest, set_cols[set_count], MAX_COLUMN_NAME,
                                           set_vals[set_count], MAX_STR_LEN);
            if (!next) break;
            set_count++;
            rest = next;
            rest = skip_spaces(rest);
        }

        if (set_count == 0) return false;

        /* Build UPDATE ... SET ... */
        size_t pos = (size_t)snprintf(sql_out, sql_size, "UPDATE %s SET ", resolved);
        for (int i = 0; i < set_count; i++) {
            if (i > 0) {
                int n = snprintf(sql_out + pos, sql_size - pos, ", ");
                if (n > 0) pos += (size_t)n;
            }
            int n = snprintf(sql_out + pos, sql_size - pos, "%s = ", set_cols[i]);
            if (n > 0) pos += (size_t)n;
            pos = append_value(sql_out, pos, sql_size, set_vals[i]);
        }

        /* Parse WHERE conditions */
        rest = skip_spaces(rest);
        if (*rest) {
            int n = snprintf(sql_out + pos, sql_size - pos, " WHERE ");
            if (n > 0) pos += (size_t)n;

            bool first = true;
            while (*rest) {
                char col[MAX_COLUMN_NAME], val[MAX_STR_LEN];
                const char *next = parse_pair(rest, col, sizeof(col), val, sizeof(val));
                if (!next) break;

                if (!first) {
                    n = snprintf(sql_out + pos, sql_size - pos, " AND ");
                    if (n > 0) pos += (size_t)n;
                }
                n = snprintf(sql_out + pos, sql_size - pos, "%s = ", col);
                if (n > 0) pos += (size_t)n;
                pos = append_value(sql_out, pos, sql_size, val);

                first = false;
                rest = next;
                rest = skip_spaces(rest);
            }
        }
        return true;
    }

    /* create <type> <name> [content] */
    if (strcasecmp(verb, "create") == 0) {
        char type[MAX_STR_LEN], name[MAX_STR_LEN];
        rest = next_token(rest, type, sizeof(type));
        if (type[0] == '\0') return false;
        rest = next_token(rest, name, sizeof(name));
        if (name[0] == '\0') return false;

        rest = skip_spaces(rest);
        if (*rest) {
            /* Content provided — strip surrounding quotes if present */
            const char *p = rest;
            if (*p == '\'') p++;
            char content[MAX_STR_LEN];
            size_t ci = 0;
            while (*p && ci < MAX_STR_LEN - 1) {
                if (*p == '\'' && *(p + 1) == '\0') break;
                content[ci++] = *p++;
            }
            content[ci] = '\0';
            snprintf(sql_out, sql_size,
                "INSERT INTO ObjectTable (name, type, data) VALUES ('%s', '%s', '%s')",
                name, type, content);
        } else {
            snprintf(sql_out, sql_size,
                "INSERT INTO ObjectTable (name, type) VALUES ('%s', '%s')",
                name, type);
        }
        return true;
    }

    /* open <name> */
    if (strcasecmp(verb, "open") == 0) {
        char name[MAX_STR_LEN];
        rest = next_token(rest, name, sizeof(name));
        if (name[0] == '\0') return false;
        snprintf(sql_out, sql_size,
            "SELECT * FROM ObjectTable WHERE name = '%s'", name);
        return true;
    }

    /* list [type] */
    if (strcasecmp(verb, "list") == 0) {
        char type[MAX_STR_LEN];
        rest = next_token(rest, type, sizeof(type));
        if (type[0] == '\0') {
            snprintf(sql_out, sql_size, "SELECT * FROM ObjectTable");
        } else {
            snprintf(sql_out, sql_size,
                "SELECT * FROM ObjectTable WHERE type = '%s'", type);
        }
        return true;
    }

    /* rm <name> */
    if (strcasecmp(verb, "rm") == 0) {
        char name[MAX_STR_LEN];
        rest = next_token(rest, name, sizeof(name));
        if (name[0] == '\0') return false;
        snprintf(sql_out, sql_size,
            "DELETE FROM ObjectTable WHERE name = '%s'", name);
        return true;
    }

    /* ps */
    if (strcasecmp(verb, "ps") == 0) {
        snprintf(sql_out, sql_size, "SELECT * FROM ProcessTable");
        return true;
    }

    /* Not a friendly command */
    return false;
}

File: vaultos-rs/kernel/src/shell/friendly.rs
// Friendly command translator: user-friendly syntax → SQL

const SQL_MAX: usize = 512;

/// Table alias mappings
static ALIASES: &[(&str, &str)] = &[
    ("procs",   "ProcessTable"),
    ("caps",    "CapabilityTable"),
    ("objects", "ObjectTable"),
    ("msgs",    "MessageTable"),
    ("audit",   "AuditTable"),
    ("config",  "SystemTable"),
    ("sys",     "SystemTable"),
];

/// Resolve alias to real table name, or return as-is.
pub fn resolve_alias(name: &str) -> &str {
    for &(alias, real) in ALIASES {
        if eq_ci(name, alias) {
            return real;
        }
    }
    name
}

/// Try to translate a friendly command to SQL. Returns Some(sql_bytes_len) on success.
pub fn translate(input: &[u8], sql: &mut [u8; SQL_MAX]) -> Option<usize> {
    let input_str = to_str(input);
    let trimmed = input_str.trim();
    if trimmed.is_empty() { return None; }

    let mut tokens = [("", 0usize, 0usize); 32]; // (token, start, end)
    let count = tokenize(trimmed, &mut tokens);
    if count == 0 { return None; }

    let verb = tokens[0].0;

    // tables / SHOW TABLES
    if eq_ci(verb, "tables") {
        return write_sql(sql, b"SHOW TABLES");
    }

    // show <table> → SELECT * FROM <table>
    if eq_ci(verb, "show") || eq_ci(verb, "list") {
        if count >= 2 {
            let table = resolve_alias(tokens[1].0);
            return write_fmt(sql, &[b"SELECT * FROM ", table.as_bytes()]);
        } else if eq_ci(verb, "list") {
            return write_sql(sql, b"SELECT * FROM ObjectTable");
        }
        return None;
    }

    // info <table> → DESCRIBE <table>
    if eq_ci(verb, "info") && count >= 2 {
        let table = resolve_alias(tokens[1].0);
        return write_fmt(sql, &[b"DESCRIBE ", table.as_bytes()]);
    }

    // count <table> → SELECT * FROM <table> (result count shown by display)
    if eq_ci(verb, "count") && count >= 2 {
        let table = resolve_alias(tokens[1].0);
        return write_fmt(sql, &[b"SELECT * FROM ", table.as_bytes()]);
    }

    // find <table> [col=val ...] → SELECT * FROM <table> WHERE ...
    if eq_ci(verb, "find") && count >= 2 {
        let table = resolve_alias(tokens[1].0);
        let mut pos = write_to(sql, 0, b"SELECT * FROM ");
        pos = write_to(sql, pos, table.as_bytes());
        if count > 2 {
            pos = write_to(sql, pos, b" WHERE ");
            pos = build_where(trimmed, &tokens, 2, count, sql, pos);
        }
        return Some(pos);
    }

    // add <table> col=val [...] → INSERT INTO <table> (...) VALUES (...)
    if eq_ci(verb, "add") && count >= 3 {
        let table = resolve_alias(tokens[1].0);
        return build_insert(trimmed, &tokens, 2, count, table, sql);
    }

    // del <table> col=val [...] → DELETE FROM <table> WHERE ...
    if eq_ci(verb, "del") || eq_ci(verb, "rm") {
        if count >= 3 {
            let table = resolve_alias(tokens[1].0);
            let mut pos = write_to(sql, 0, b"DELETE FROM ");
            pos = write_to(sql, pos, table.as_bytes());
            pos = write_to(sql, pos, b" WHERE ");
            pos = build_where(trimmed, &tokens, 2, count, sql, pos);
            return Some(pos);
        }
        // rm <name> → DELETE FROM ObjectTable WHERE name = '<name>'
        if eq_ci(verb, "rm") && count >= 2 {
            let name = tokens[1].0;
            let mut pos = write_to(sql, 0, b"DELETE FROM ObjectTable WHERE name = '");
            pos = write_escaped(sql, pos, name.as_bytes());
            pos = write_to(sql, pos, b"'");
            return Some(pos);
        }
        return None;
    }

    // set <table> col=val [...] where k=v → UPDATE <table> SET ... WHERE ...
    if eq_ci(verb, "set") && count >= 4 {
        let table = resolve_alias(tokens[1].0);
        return build_update(trimmed, &tokens, 2, count, table, sql);
    }

    // create <type> <name> [content] → INSERT INTO ObjectTable
    if eq_ci(verb, "create") && count >= 3 {
        let obj_type = tokens[1].0;
        let name = tokens[2].0;
        let mut pos = write_to(sql, 0, b"INSERT INTO ObjectTable (name, type");
        if count > 3 {
            pos = write_to(sql, pos, b", data) VALUES ('");
        } else {
            pos = write_to(sql, pos, b") VALUES ('");
        }
        pos = write_escaped(sql, pos, name.as_bytes());
        pos = write_to(sql, pos, b"', '");
        pos = write_escaped(sql, pos, obj_type.as_bytes());
        pos = write_to(sql, pos, b"'");
        if count > 3 {
            // Rest of line is content
            let content_start = tokens[3].1;
            let content = &trimmed[content_start..];
            pos = write_to(sql, pos, b", '");
            pos = write_escaped(sql, pos, content.as_bytes());
            pos = write_to(sql, pos, b"'");
        }
        pos = write_to(sql, pos, b")");
        return Some(pos);
    }

    // open <name> → SELECT * FROM ObjectTable WHERE name = '<name>'
    if eq_ci(verb, "open") || eq_ci(verb, "cat") {
        if count >= 2 {
            let name = tokens[1].0;
            let mut pos = write_to(sql, 0, b"SELECT * FROM ObjectTable WHERE name = '");
            pos = write_escaped(sql, pos, name.as_bytes());
            pos = write_to(sql, pos, b"'");
            return Some(pos);
        }
        return None;
    }

    // ps → SELECT * FROM ProcessTable
    if eq_ci(verb, "ps") {
        return write_sql(sql, b"SELECT * FROM ProcessTable");
    }

    None // Not a friendly command; try as raw SQL
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

fn build_where(
    _input: &str, tokens: &[(&str, usize, usize); 32],
    start: usize, count: usize, sql: &mut [u8; SQL_MAX], mut pos: usize,
) -> usize {
    let mut first = true;
    for i in start..count {
        let tok = tokens[i].0;
        if eq_ci(tok, "where") { continue; }
        if let Some(eq_pos) = tok.find('=') {
            if !first { pos = write_to(sql, pos, b" AND "); }
            let col = &tok[..eq_pos];
            let val = &tok[eq_pos + 1..];
            pos = write_to(sql, pos, col.as_bytes());
            pos = write_to(sql, pos, b" = ");
            pos = append_value(sql, pos, val);
            first = false;
        }
    }
    pos
}

fn build_insert(
    _input: &str, tokens: &[(&str, usize, usize); 32],
    start: usize, count: usize, table: &str, sql: &mut [u8; SQL_MAX],
) -> Option<usize> {
    let mut cols: [&str; 16] = [""; 16];
    let mut vals: [&str; 16] = [""; 16];
    let mut n = 0;

    for i in start..count {
        let tok = tokens[i].0;
        if let Some(eq_pos) = tok.find('=') {
            if n >= 16 { break; }
            cols[n] = &tok[..eq_pos];
            vals[n] = &tok[eq_pos + 1..];
            n += 1;
        }
    }
    if n == 0 { return None; }

    let mut pos = write_to(sql, 0, b"INSERT INTO ");
    pos = write_to(sql, pos, table.as_bytes());
    pos = write_to(sql, pos, b" (");
    for i in 0..n {
        if i > 0 { pos = write_to(sql, pos, b", "); }
        pos = write_to(sql, pos, cols[i].as_bytes());
    }
    pos = write_to(sql, pos, b") VALUES (");
    for i in 0..n {
        if i > 0 { pos = write_to(sql, pos, b", "); }
        pos = append_value(sql, pos, vals[i]);
    }
    pos = write_to(sql, pos, b")");
    Some(pos)
}

fn build_update(
    _input: &str, tokens: &[(&str, usize, usize); 32],
    start: usize, count: usize, table: &str, sql: &mut [u8; SQL_MAX],
) -> Option<usize> {
    // Find "where" keyword
    let mut where_idx = count;
    for i in start..count {
        if eq_ci(tokens[i].0, "where") {
            where_idx = i;
            break;
        }
    }

    let mut pos = write_to(sql, 0, b"UPDATE ");
    pos = write_to(sql, pos, table.as_bytes());
    pos = write_to(sql, pos, b" SET ");

    let mut first = true;
    for i in start..where_idx {
        let tok = tokens[i].0;
        if let Some(eq_pos) = tok.find('=') {
            if !first { pos = write_to(sql, pos, b", "); }
            let col = &tok[..eq_pos];
            let val = &tok[eq_pos + 1..];
            pos = write_to(sql, pos, col.as_bytes());
            pos = write_to(sql, pos, b" = ");
            pos = append_value(sql, pos, val);
            first = false;
        }
    }

    if where_idx < count {
        pos = write_to(sql, pos, b" WHERE ");
        pos = build_where(_input, tokens, where_idx + 1, count, sql, pos);
    }

    Some(pos)
}

/// Escape single quotes in user input for safe SQL embedding (' → '').
fn write_escaped(sql: &mut [u8; SQL_MAX], pos: usize, input: &[u8]) -> usize {
    let mut p = pos;
    for &b in input {
        if p >= SQL_MAX - 1 { break; }
        if b == b'\'' {
            p = write_to(sql, p, b"''");
        } else {
            sql[p] = b;
            sql[p + 1] = 0;
            p += 1;
        }
    }
    p
}

fn append_value(sql: &mut [u8; SQL_MAX], mut pos: usize, val: &str) -> usize {
    if is_numeric(val) || eq_ci(val, "true") || eq_ci(val, "false") {
        pos = write_to(sql, pos, val.as_bytes());
    } else {
        pos = write_to(sql, pos, b"'");
        pos = write_escaped(sql, pos, val.as_bytes());
        pos = write_to(sql, pos, b"'");
    }
    pos
}

fn is_numeric(s: &str) -> bool {
    if s.is_empty() { return false; }
    let bytes = s.as_bytes();
    let start = if bytes[0] == b'-' { 1 } else { 0 };
    if start >= bytes.len() { return false; }
    bytes[start..].iter().all(|&b| b >= b'0' && b <= b'9')
}

fn tokenize<'a>(input: &'a str, tokens: &mut [(&'a str, usize, usize); 32]) -> usize {
    let mut count = 0;
    let mut i = 0;
    let bytes = input.as_bytes();
    let len = bytes.len();

    while i < len && count < 32 {
        // Skip whitespace
        while i < len && bytes[i] == b' ' { i += 1; }
        if i >= len { break; }

        let start = i;
        // Handle quoted strings as single token
        if bytes[i] == b'\'' || bytes[i] == b'"' {
            let quote = bytes[i];
            i += 1;
            while i < len && bytes[i] != quote { i += 1; }
            if i < len { i += 1; }
        } else {
            while i < len && bytes[i] != b' ' { i += 1; }
        }
        tokens[count] = (&input[start..i], start, i);
        count += 1;
    }
    count
}

fn eq_ci(a: &str, b: &str) -> bool {
    if a.len() != b.len() { return false; }
    a.as_bytes().iter().zip(b.as_bytes()).all(|(&x, &y)| {
        x.to_ascii_lowercase() == y.to_ascii_lowercase()
    })
}

fn write_to(sql: &mut [u8; SQL_MAX], pos: usize, data: &[u8]) -> usize {
    if pos >= SQL_MAX - 1 { return pos; }
    let avail = SQL_MAX - pos - 1;
    let len = if data.len() > avail { avail } else { data.len() };
    sql[pos..pos + len].copy_from_slice(&data[..len]);
    sql[pos + len] = 0;
    pos + len
}

fn write_sql(sql: &mut [u8; SQL_MAX], data: &[u8]) -> Option<usize> {
    let pos = write_to(sql, 0, data);
    Some(pos)
}

fn write_fmt(sql: &mut [u8; SQL_MAX], parts: &[&[u8]]) -> Option<usize> {
    let mut pos = 0;
    for part in parts {
        pos = write_to(sql, pos, part);
    }
    Some(pos)
}

fn to_str(bytes: &[u8]) -> &str {
    let len = bytes.iter().position(|&b| b == 0).unwrap_or(bytes.len());
    core::str::from_utf8(&bytes[..len]).unwrap_or("")
}
